reaper/repair_run_test.go

```diff
@@ -0,0 +1,407 @@
+		assert.NotNil(t, segment.StartTime)
+		assert.NotNil(t, segment.EndTime)
+		checkRepairRunSegment(t, run, segment)
+	}
+}
+
+func testAbortRepairRunSegments(t *testing.T, client Client) {
+	run := createRepairRun(t, client, "cluster-1")
+	defer deleteRepairRun(t, client, run)
+	err := client.StartRepairRun(context.Background(), run.Id)
+	require.Nil(t, err)
+	segments := waitForSegmentsStarted(t, client, run)
+	for _, segment := range segments {
+		// could be STARTED, RUNNING or DONE
+		assert.NotEqual(t, RepairSegmentStateNotStarted, segment.State)
+		err = client.AbortRepairRunSegment(context.Background(), run.Id, segment.Id)
+		if err != nil {
+			require.Contains(t, err.Error(), "Cannot abort segment on repair run with status DONE")
+		}
+	}
+	segments, err = client.GetRepairRunSegments(context.Background(), run.Id)
+	require.Nil(t, err)
+	require.Len(t, segments, 2)
+	for _, segment := range segments {
+		assert.True(t,
+			segment.State == RepairSegmentStateNotStarted ||
+				segment.State == RepairSegmentStateDone)
+	}
+}
+
+func testDeleteRepairRunNotFound(t *testing.T, client Client) {
+	nonExistentRepairRun, _ := uuid.NewUUID()
+	err := client.DeleteRepairRun(context.Background(), nonExistentRepairRun, "Alice")
+	assert.NotNil(t, err)
+	// Reaper returns a spurious '%s' in the error message
+	assert.Contains(t, err.Error(), fmt.Sprintf("Repair run %%s%v not found", nonExistentRepairRun))
+}
+
+func testPurgeRepairRun(t *testing.T, client Client) {
+	run := createRepairRun(t, client, "cluster-1")
+	defer deleteRepairRun(t, client, run)
+	purged, err := client.PurgeRepairRuns(context.Background())
+	require.Nil(t, err)
+	assert.Equal(t, 0, purged)
+}
+
+func createRepairRun(t *testing.T, client Client, cluster string) *RepairRun {
+	runId, err := client.CreateRepairRun(
+		context.Background(),
+		cluster,
+		keyspace,
+		"Alice",
+		&RepairRunCreateOptions{
+			Tables:              []string{"table1", "table2"},
+			SegmentCountPerNode: 1,
+			RepairParallelism:   RepairParallelismParallel,
+			Intensity:           0.1,
+			IncrementalRepair:   true,
+			RepairThreadCount:   4,
+			Cause:               "testing repair runs",
+		},
+	)
+	require.Nil(t, err)
+	repairRun, err := client.GetRepairRun(context.Background(), runId)
+	require.Nil(t, err)
+	return checkRepairRun(t, cluster, repairRun)
+}
+
+func checkRepairRun(t *testing.T, cluster string, actual *RepairRun) *RepairRun {
+	assert.NotNil(t, actual.Id)
+	assert.Equal(t, cluster, actual.Cluster)
+	assert.Equal(t, keyspace, actual.Keyspace)
+	assert.Equal(t, "Alice", actual.Owner)
+	assert.Equal(t, "testing repair runs", actual.Cause)
+	assert.ElementsMatch(t, []string{"table1", "table2"}, actual.Tables)
+	assert.Equal(t, RepairRunStateNotStarted, actual.State)
+	assert.InDelta(t, 0.1, actual.Intensity, 0.001)
+	assert.Equal(t, true, actual.IncrementalRepair)
+	// FIXME segment count per node is always 2
+	// assert.Equal(t, 1, actual.SegmentCountPerNode)
```

Comment:
I'm seeing a strange behavior here. The segment count per node is always 2, no matter which value is given when creating the repair run. @adejanovski do you know what could be happening here?

Reply:
Also, I'm calling this field and similar ones `SegmentCountPerNode` to harmonize names. The actual JSON field is called `segmentCount` or `segmentCountPerNode` in requests, and `total_segments` in responses. From what I can infer by reading Reaper's code, they all represent the same notion.

Reply:
This is indeed a little confusing, so I hopefully can make things clearer:
Initially Reaper was requesting a total number of segments whatever the cluster size when defining repair schedules and repair runs. This didn't scale well and required to adjust the number of segments over time as the cluster scaled.
We moved to a number of segments per node later on but maintained backwards compatibility for a while, which is why you can have both names in the schedules definition.
One thing to note is that "segment count per node" is a notion that exists on schedules and when creating a repair run. Repair runs themselves don't store this field, they have the actual segments that were generated based on the requested segment count per node, and the `total_segments` value that you see in the response which is the count of segments that were generated for the run.

Incremental repairs are special because we're generating "fake" segments. Incremental repair doesn't support subrange, so the repair technique is to run it once per node on all ranges. We're then generating one segment per node in the cluster and will not consider the segment count per node value. I guess that's why you're getting 2 segments consistently in your tests (that shouldn't happen with full repairs though).

Reply:
Thank you @adejanovski, this is very helpful ! I think I should update this PR and rename `RepairRun.SegmentCountPerNode` to `RepairRun.TotalSegments` as per your explanations. I will also update the tests and try a non-incremental repair to check if I get the expected amount of segments.

Reply:
Implemented in eb34b5c. It seems it is still not possible to guess the exact number of segments, but at least I'm getting bigger numbers.

Reply:
right, there will always be a slightly different number of segments than what you expect due to Reaper trying its best to generate equally sized segments.